api/turing/service/ensembling_job_service.go

```diff
@@ -17,7 +21,8 @@ const (
 	SparkHomeFolder = "/home/spark"
 	// EnsemblerFolder is the folder created by the Turing SDK that contains
 	// the ensembler dependencies and pickled Python files.
-	EnsemblerFolder = "ensembler"
+	EnsemblerFolder  = "ensembler"
+	jobNameMaxLength = 20
```

Comment:
Ohh why this limit? I'm afraid having this constraint is not very generalisable. Is it a limitation from the internal dashboard?

Reply:
It's a limit because of how spark truncates the pod name, I don't remember the exact value because I used trial and error, the value could be more but 20 seems safe. Are you suggesting hardcoding it?

Reply:
I think there are certain implications that were overlooked here as discussed with @krithika369 offline.

This setting is reversed engineering from how spark generates the pod names. Generally speaking, there are a few components at play here and I will try my best to explain what is generated in terms of order.

1. Ensembling job name provided by the user is appended with the unix timestamp appended to the back. This is currently limited to 50 characters excluding the timestamp (60 included).
2. SparkApplication (CRD) is created with the name equivalent to ensembling job name which has a maximum of 60 characters.
3. Spark Driver Pod is created with the Pod name of a format roughly equivalent to this: `<Ensembling Job Name truncated to some length>-<generated hash>-driver`
4. Spark Executor Pod is created with the Pod name of a format roughly equivalent to this: `<Ensembling Job Name truncated to some length>-<generated hash>-exec-<replica number>`

So the dashboard works by using the metric provided by kube-state-metrics: `kube_pod_container_status_ready` and other relevant metric names by filtering it based on `pod=~"$job.*-driver|$job.*-exec-.*"`. According to the [kube-state-metrics documentation](https://github.com/kubernetes/kube-state-metrics/blob/master/docs/pod-metrics.md) there seems to be no filter based on a pod annotation/label so this becomes extremely awkward and the reason why I've decided to do a string operation to extract the value out of it.

Now the problem is how are we expecting our users to name their ensembling jobs? @krithika369 probably suspects that they are going to create a name based on a fixed portion appended by some generated string; i.e. in the format `<fixed name>-<generated>` which might be a problem if we truncate to such short character sets of length 20.

My proposal is to limit the number of characters in the CreateEnsemblingJob process to 20 characters. This probably makes things a little tough. If we append 10 characters for the timestamp, users are basically left with 10 characters for their ensembling job name. A few approaches I can think of:

1. Recommend them to use a name that is short and tell them there will be a timestamp generated at the end of it, limiting them to 10 characters only. 
2. Leave the responsibility of letting them generate a name themselves.

In both circumstances, we would have to enforce a unique index on the name + mlp project or there might be collisions in naming.

Honestly speaking, I'm not sure what should be the right way to approach this problem. Maybe @romanwozniak you could weigh in on this.

Reply:
> My proposal is to limit the number of characters in the CreateEnsemblingJob process to 20 characters

After thinking about this a bit and discussing with @ashwinath offline, limiting the Ensembler name length sounds like a reasonable approach to me (for comparison, we set a 25 chart limit for [router names](https://github.com/gojek/turing/blob/main/ui/src/router/components/form/validation/schema.js#L214) which is probably the reason why we don't have to worry about any truncation of the k8s resource name later). We could put it the same constraint for ensembler names. And similar to the router version number, we can introduce the concept of a `run-id` in the jobs table that is unique per `(MLP Project ID, Ensembler ID)` so that the final name of the job (ensembler name + run ID, instead of name + timestamp) is still < 30 characters and allows enough room for the other suffixes appended by the Spark op on its resources.

> In both circumstances, we would have to enforce a unique index on the name + mlp project or there might be collisions in naming.

Makes sense to me. We do a unique constraint on (MLP Project ID, Router Name). I think the same logic would work for Ensemblers - (MLP Project ID, Ensembler Name).

Reply:
This also affects the ensembler name as we create the ensembling job name based on the ensembler name.

Reply:
RunID approach, proposed by @krithika369 makes sense to me. So then we can limit ensembler's name length to 25 char (to make it consistent with routers) and append sequential run id. But propose to implement it in a separate PR